proposals/NNNN-allow-taskgroup-childtaskresult-type-to-be-inferred.md

```diff
@@ -0,0 +1,138 @@
+
+```swift
+let s = withTaskGroup { group in
+  group.addTask { /* ... */ }
+  group.addTask { /* ... */ }
+  group.addTask { /* ... */ }
+
+  return "Hello, world!"
+}
+```
+
+In the above snippet, `ChildTaskResult` is inferred as `Void` and `GroupResult` is inferred as `String`.  Not needing to specify the generics explicitly will simplify the API design for these functions and make it easier for new users of these APIs, as it can currently be confusing to understand the differences between `ChildTaskResult` and `GroupResult` (especially when one or both of those is `Void`).
+
+## Detailed design
+
+Because [SE-0326](0326-extending-multi-statement-closure-inference.md) will look at the first statement to determine the generic for `ChildTaskResult`, it is possible to get a compiler error by creating a task group like so:
+
+```swift
+// Expect `ChildTaskResult` to be `Void`...
+withTaskGroup { group in // generic parameter 'ChildTaskResult' could not be inferred
+    // Since `addTask` wasn't the first statement, this fails to compile.
+    group.cancelAll()
+    group.addTask { /* ... */ }
+}
+```
+
+This can be fixed by going back to specifying the generic like before:
+
+```swift
+// Expect `ChildTaskResult` to be `Void`...
+withTaskGroup(of: Void.self) { group in
+    group.cancelAll()
+    group.addTask { /* ... */ }
+}
+```
+
+However, this is a rare case in general since `addTask` is generally the first statement in a task group body.
+
+It is also possible to create a compiler error by returning two different values from a `addTask` closure:
+
+```swift
+withTaskGroup { group in
+    group.addTask { "Hello, world!" }
+    group.addTask { 42 } // Cannot convert value of type 'Int' to closure result type 'String'
+}
+```
+
+The compiler will already give a good error message here, since the first `addTask` statement is what determined (in this case) that the `ChildTaskResult` generic was set to `String`.  If this needs to be made more clear (instead of being inferred), the user can always specify the generic directly as before:
+
+```swift
+withTaskGroup(of: Int.self) { group in
+    // Now the error has moved here since the generic was specified up front...
+    group.addTask { "Hello, world!" } // Cannot convert value of type 'String' to closure result type 'Int'
+    group.addTask { 42 }
+}
+```
+
+## Source compatibility
+
+Omitting the `of childTaskResultType: ChildTaskResult.Type` parameter for both `withTaskGroup(of:returning:body:)` and `withThrowingTaskGroup(of:returning:body:)` is new, and therefore the inference of `ChildTaskResult` is opt-in and does not break source compatibility.
+
+## ABI compatibility
+
+No ABI impact since this is an additive change to default arguments.
+
+## Implications on adoption
+
+This feature can be freely adopted and un-adopted in source
+code with no deployment constraints and without affecting source or ABI
+compatibility.
+
+## Future directions
+
+...
```

Comment:
Maybe let's add this:

---


### TaskGroup APIs without the "with..." closures

While not possible without more compiler features to enforce the safety of a task group not escaping a context, and having to await all of its results at the end of a scope... it is an interesting future direction to explore a TaskGroup API that does not need to resort to `with...` methods, like this:

```
// Potential long-term direction that might be possible
func test() async { 
  let group = TaskGroup<Int>()
  group.addTask { ... }
  
  // going out of scope would have to imply group.awaitAll()
}
```

If we were to explore such API, the type inference rules would be somewhat different, and a TaskGroup would likely be initialized more similarly to a collection: `TaskGroup<Int>`.

This proposal has no impact on this future direction, and can be accepted as is, without precluding future developments in API ergonomics like this.